server/src/main/java/org/elasticsearch/indices/TimestampFieldMapperService.java

```diff
@@ -0,0 +1,174 @@
+
+                final IndexService indexService = indicesService.indexService(index);
+                if (indexService == null) {
+                    logger.trace("computing timestamp mapping for {} async", index);
+                    executor.execute(new AbstractRunnable() {
+                        @Override
+                        public void onFailure(Exception e) {
+                            logger.debug(new ParameterizedMessage("failed to compute mapping for {}", index), e);
+                            future.onResponse(null); // no need to propagate a failure to create the mapper service to searches
+                        }
+
+                        @Override
+                        protected void doRun() throws Exception {
+                            try (MapperService mapperService = indicesService.createIndexMapperService(indexMetadata)) {
+                                mapperService.merge(indexMetadata, MapperService.MergeReason.MAPPING_RECOVERY);
+                                future.onResponse(fromMapperService(mapperService));
+                            }
+                        }
+                    });
+                } else {
+                    logger.trace("computing timestamp mapping for {} using existing index service", index);
+                    try {
+                        future.onResponse(fromMapperService(indexService.mapperService()));
+                    } catch (Exception e) {
+                        assert false : e;
+                        future.onResponse(null);
+                    }
+                }
+            }
+        }
+    }
+
+    private static boolean hasUsefulTimestampField(IndexMetadata indexMetadata) {
+        if (indexMetadata == null) {
+            return false;
+        }
+        final IndexLongFieldRange timestampMillisRange = indexMetadata.getTimestampMillisRange();
+        return timestampMillisRange.isComplete() && timestampMillisRange != IndexLongFieldRange.UNKNOWN;
+    }
+
+    private static DateFieldMapper.DateFieldType fromMapperService(MapperService mapperService) {
+        final MappedFieldType mappedFieldType = mapperService.fieldType(DataStream.TimestampField.FIXED_TIMESTAMP_FIELD);
+        if (mappedFieldType instanceof DateFieldMapper.DateFieldType) {
+            return (DateFieldMapper.DateFieldType) mappedFieldType;
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * @return the field type of the {@code @timestamp} field of the given index, or {@code null} if:
+     * - the index is not found,
+     * - the field is not found,
+     * - the mapping is not known yet, or
+     * - the field is not a timestamp field.
+     */
+    @Nullable
+    public DateFieldMapper.DateFieldType getTimestampFieldType(Index index) {
+        final PlainActionFuture<DateFieldMapper.DateFieldType> future = fieldTypesByIndex.get(index);
+        if (future == null || future.isDone() == false) {
```

Comment:
There remains a question of whether we should block here or not (and if so, for how long).

On reflection I think we shouldn't block. Returning `null` sooner will allow the search coordination to proceed normally, ignoring any timestamp filter and deferring any skipping to the individual shards. This means we'll see shard failures if the coordinating node falls behind on extracting these mappings AND some of the shards are unassigned, which is hopefully rare.

As a follow-up we could in theory add another more patient getter to support a workflow that goes:

1. we call `getTimestampFieldType` which returns `null`
2. some shards are unavailable for the `can_match` phase
3. we call `getTimestampFieldTypePatiently` to see for whether those shard failures can be ignored or not

Reply:
Right, although in that case there's no particular reason to expect shards to be unavailable.